test/polygon/hop/hop.test.ts

```diff
@@ -0,0 +1,240 @@
+  const DAI_ADDR = "0x8f3cf7ad23cd3cadbd9735aff958023239c6a063";
+  const hDai = "0xb8901acB165ed027E32754E0FFe830802919727f";
+  const l2AmmWrapper = "0x28529fec439cfF6d7D1D5917e956dEE62Cd3BE5c";
+  const l2Bridge = "0xEcf268Be00308980B5b3fcd0975D47C4C8e1382a";
+  const saddleSwap = "0x25FB92E505F752F730cAD0Bd4fa17ecE4A384266";
+  const saddleSwapInstance = new ethers.Contract(saddleSwap, saddleSwapABI);
+
+  const token = new ethers.Contract(DAI_ADDR, abis.basic.erc20);
+  const l2BridgeInstance = new ethers.Contract(l2Bridge, l2BridgeABI);
+
+  before(async () => {
+    await hre.network.provider.request({
+      method: "hardhat_reset",
+      params: [
+        {
+          forking: {
+            // @ts-ignore
+            jsonRpcUrl: hre.config.networks.hardhat.forking.url,
+            blockNumber: 27054896
+          }
+        }
+      ]
+    });
+
+    masterSigner = await getMasterSigner();
+    instaConnectorsV2 = await ethers.getContractAt(abis.core.connectorsV2, addresses.core.connectorsV2);
+    connector = await deployAndEnableConnector({
+      connectorName,
+      contractArtifact: ConnectV2HopPolygon__factory,
+      signer: masterSigner,
+      connectors: instaConnectorsV2
+    });
+    console.log("Connector address", connector.address);
+  });
+
+  it("Should have contracts deployed.", async function () {
+    expect(!!instaConnectorsV2.address).to.be.true;
+    expect(!!connector.address).to.be.true;
+    expect(!!(await masterSigner.getAddress())).to.be.true;
+  });
+
+  describe("DSA wallet setup", function () {
+    it("Should build DSA v2", async function () {
+      dsaWallet0 = await buildDSAv2(wallet0.address);
+      expect(!!dsaWallet0.address).to.be.true;
+    });
+
+    it("Deposit MATIC & DAI into DSA wallet", async function () {
+      await wallet0.sendTransaction({
+        to: dsaWallet0.address,
+        value: ethers.utils.parseEther("10")
+      });
+      expect(await ethers.provider.getBalance(dsaWallet0.address)).to.be.gte(ethers.utils.parseEther("10"));
+
+      await addLiquidity("dai", dsaWallet0.address, ethers.utils.parseEther("10000"));
+    });
+  });
+
+  describe("Main", function () {
+    it("should migrate from L2 to L1", async function () {
+      const amount = ethers.utils.parseEther("10");
+      const bonderFee = new BigNumber(100 * 1e9)
+        .multipliedBy(150000)
+        .multipliedBy(1.5)
+        .plus(new BigNumber(amount.toString()).multipliedBy(0.18));
+      const deadline = Date.now() + 604800;
+      const getId = "0";
+
+      await token.connect(wallet0).approve(l2AmmWrapper, amount.toString());
+
+      const hDaiIdx = await saddleSwapInstance.connect(wallet0).getTokenIndex(hDai);
+      const daiIdx = await saddleSwapInstance.connect(wallet0).getTokenIndex(DAI_ADDR);
+
+      let amountOutMin = await saddleSwapInstance.connect(wallet0).calculateSwap(daiIdx, hDaiIdx, amount.toString());
+      amountOutMin = new BigNumber(amountOutMin.toString()).multipliedBy(0.99);
```

Comment:
@shanefontaine This is correct?

Reply:
That appears correct. The Hop SDK makes those same calculations [here](https://github.com/hop-protocol/hop/blob/develop/packages/sdk/src/AMM.ts#L288-L303).

Reply:
Okay thanks